File: common/proto/object/datasource.go
/*
 * Copyright (c) 2019-2021. Abstrium SAS <team (at) pydio.com>
 * This file is part of Pydio Cells.
 *
 * Pydio Cells is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pydio Cells is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pydio Cells.  If not, see <http://www.gnu.org/licenses/>.
 *
 * The latest code can be found at <https://pydio.com>.
 */

package object

import (
	"fmt"
	"strings"

	"go.uber.org/zap/zapcore"

	"github.com/pydio/cells/v4/common/proto/service"
	"github.com/pydio/cells/v4/common/utils/configx"
)

const (
	StorageKeyFolder       = "folder"
	StorageKeyFolderCreate = "create"
	StorageKeyNormalize    = "normalize"

	StorageKeyCustomEndpoint   = "customEndpoint"
	StorageKeyCustomRegion     = "customRegion"
	StorageKeyMinioServer      = "minioServer"
	StorageKeyBucketsTags      = "bucketsTags"
	StorageKeyObjectsTags      = "objectsTags"
	StorageKeyNativeEtags      = "nativeEtags"
	StorageKeyBucketsRegexp    = "bucketsRegexp"
	StorageKeyReadonly         = "readOnly"
	StorageKeyJsonCredentials  = "jsonCredentials"
	StorageKeyStorageClass     = "storageClass"
	StorageKeySignatureVersion = "signatureVersion"

	StorageKeyCellsInternal    = "cellsInternal"
	StorageKeyInitFromBucket   = "initFromBucket"
	StorageKeyInitFromSnapshot = "initFromSnapshot"

	AmazonS3Endpoint = "s3.amazonaws.com"
)

func (d *DataSource) ClientConfig() configx.Values {
	cfg := configx.New()
	_ = cfg.Val("type").Set("mc")
	if d.StorageType == StorageType_AZURE {
		_ = cfg.Val("type").Set("azure")
	}
	_ = cfg.Val("endpoint").Set(d.BuildUrl())
	_ = cfg.Val("key").Set(d.GetApiKey())
	_ = cfg.Val("secret").Set(d.GetApiSecret())
	_ = cfg.Val("secure").Set(d.GetObjectsSecure())
	_ = cfg.Val("minioServer").Set(d.ServerIsMinio())
	if d.StorageConfiguration != nil {
		if r, o := d.StorageConfiguration[StorageKeyCustomRegion]; o && r != "" {
			_ = cfg.Val("region").Set(r)
		}
		if ce, o := d.StorageConfiguration[StorageKeyCustomEndpoint]; o && ce != "" {
			_ = cfg.Val(StorageKeyCustomEndpoint).Set(ce)
		}
		if sv, o := d.StorageConfiguration[StorageKeySignatureVersion]; o && sv != "" {
			_ = cfg.Val("signature").Set(sv)
		}
	}
	return cfg
}

// BuildUrl constructs the url used for clients.
func (d *DataSource) BuildUrl() string {
	host := d.ObjectsHost
	if d.ObjectsHost == "" {
		host = "localhost"
	}
	if d.ObjectsPort == 443 || d.ObjectsPort == 0 {
		return host
	} else {
		return fmt.Sprintf("%s:%d", host, d.ObjectsPort)
	}
}

// ServerIsMinio checks if server is local, of if remote definition has the minioServer key
func (d *DataSource) ServerIsMinio() bool {
	if d.StorageType == StorageType_LOCAL {
		return true
	}
	if d.StorageConfiguration == nil {
		return false
	}
	if m, o := d.StorageConfiguration[StorageKeyMinioServer]; o && m == "true" {
		return true
	}
	return false
}

func (d *MinioConfig) ClientConfig() configx.Values {
	cfg := configx.New()
	_ = cfg.Val("type").Set("mc")
	_ = cfg.Val("endpoint").Set(d.BuildUrl())
	_ = cfg.Val("key").Set(d.GetApiKey())
	_ = cfg.Val("secret").Set(d.GetApiSecret())
	_ = cfg.Val("secure").Set(d.GetRunningSecure())
	if d.StorageType == StorageType_LOCAL {
		_ = cfg.Val("minioServer").Set(true)
	} else if d.StorageType == StorageType_AZURE {
		_ = cfg.Val("type").Set("azure")
	}
	if d.GatewayConfiguration != nil {
		if m, o := d.GatewayConfiguration[StorageKeyMinioServer]; o && m == "true" {
			_ = cfg.Val("minioServer").Set(true)
		}
		if r, o := d.GatewayConfiguration[StorageKeyCustomRegion]; o && r != "" {
			_ = cfg.Val("region").Set(r)
		}
		if ce, o := d.GatewayConfiguration[StorageKeyCustomEndpoint]; o && ce != "" {
			_ = cfg.Val("customEndpoint").Set(ce)
		}
		if sv, o := d.GatewayConfiguration[StorageKeySignatureVersion]; o && sv != "" {
			_ = cfg.Val("signature").Set(sv)
		}
	}
	return cfg
}

// BuildUrl builds the url used for clients
func (d *MinioConfig) BuildUrl() string {
	host := d.RunningHost
	if d.RunningHost == "" {
		host = "localhost"
	}
	if d.RunningPort == 443 || d.RunningPort == 0 {
		return host
	} else {
		return fmt.Sprintf("%s:%d", host, d.RunningPort)
	}
}

// IsInternal is a shorthand to check StorageConfiguration["cellsInternal"] key
func (d *DataSource) IsInternal() bool {
	if d.StorageConfiguration != nil {
		_, i := d.StorageConfiguration[StorageKeyCellsInternal]
		return i
	}
	return false
}

/* LOGGING SUPPORT */

// MarshalLogObject implements custom marshalling for datasource, to avoid logging ApiKey
func (d *DataSource) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
	if d == nil {
		return nil
	}
	if d.Name != "" {
		encoder.AddString("Name", d.Name)
	}
	if d.ObjectsHost != "" {
		if d.ObjectsPort != 0 {
			encoder.AddString("Host", fmt.Sprintf("%s:%d", d.ObjectsHost, d.ObjectsPort))
		} else {
			encoder.AddString("Host", d.ObjectsHost)
		}
	}
	if d.ObjectsBucket != "" {
		encoder.AddString("Bucket", d.ObjectsBucket)
	}
	encoder.AddString("StorageType", d.StorageType.String())
	if d.PeerAddress != "" {
		encoder.AddString("PeerAddress", d.PeerAddress)
	}
	if d.EncryptionMode != EncryptionMode_CLEAR {
		encoder.AddString("EncryptionMode", d.EncryptionMode.String())
	}
	return nil
}

// MarshalLogObject implements custom marshalling for Minio, to avoid logging ApiKey
func (d *MinioConfig) MarshalLogObject(encoder zapcore.ObjectEncoder) error {
	if d == nil {
		return nil
	}
	if d.Name != "" {
		encoder.AddString("Name", d.Name)
	}
	if d.RunningHost != "" {
		if d.RunningPort != 0 {
			encoder.AddString("Host", fmt.Sprintf("%s:%d", d.RunningHost, d.RunningPort))
		} else {
			encoder.AddString("Host", d.RunningHost)
		}
	}
	if d.RunningSecure {
		encoder.AddBool("Secure", d.RunningSecure)
	}
	encoder.AddString("StorageType", d.StorageType.String())
	if d.PeerAddress != "" {
		encoder.AddString("PeerAddress", d.PeerAddress)
	}
	return nil
}

func (m *DataSourceSingleQuery) Matches(object interface{}) bool {
	ds, ok := object.(*DataSource)
	if !ok {
		return false
	}
	var bb []bool
	if m.Name != "" {
		bb = append(bb, compareStrings(ds.Name, m.Name))
	}
	if m.ObjectServiceName != "" {
		bb = append(bb, compareStrings(ds.ObjectsServiceName, m.ObjectServiceName))
	}
	if m.StorageType != StorageTypeFilter_ANY {
		if m.StorageType == StorageTypeFilter_LOCALFS {
			bb = append(bb, ds.StorageType == StorageType_LOCAL)
		} else {
			bb = append(bb, ds.StorageType != StorageType_LOCAL)
		}
	}
	if m.IsDisabled {
		bb = append(bb, ds.Disabled)
	}
	// Check versioning
	if m.IsVersioned {
		bb = append(bb, ds.VersioningPolicyName != "")
	}
	if m.VersioningPolicyName != "" {
		bb = append(bb, compareStrings(ds.VersioningPolicyName, m.VersioningPolicyName))
	}
	// Check encryption
	if m.IsEncrypted {
		bb = append(bb, ds.EncryptionMode != EncryptionMode_CLEAR)
	}
	if m.EncryptionMode != EncryptionMode_CLEAR {
		bb = append(bb, m.EncryptionMode == m.EncryptionMode)
	}
	if m.EncryptionKey != "" {
		bb = append(bb, compareStrings(ds.EncryptionKey, m.EncryptionKey))
	}
	if m.FlatStorage {
		bb = append(bb, ds.FlatStorage)
	}
	if m.SkipSyncOnRestart {
		bb = append(bb, ds.SkipSyncOnRestart)
	}
	if m.PeerAddress != "" {
		bb = append(bb, compareStrings(ds.PeerAddress, m.PeerAddress))
	}
	if m.StorageConfigurationName != "" {
		cf := ds.StorageConfiguration
		if cf == nil {
			cf = map[string]string{}
		}
		val, ok := ds.StorageConfiguration[m.StorageConfigurationName]
		if m.StorageConfigurationValue == "" {
			bb = append(bb, ok)
		} else {
			bb = append(bb, compareStrings(val, m.StorageConfigurationValue))
		}
	}

	result := service.ReduceQueryBooleans(bb, service.OperationType_AND)
	if m.Not {
		return !result
	} else {
		return result
	}
}

func compareStrings(ref, search string) bool {
	// Basic search: can have wildcard on left, right, or none (exact search)
	var left, right bool
	if strings.HasPrefix(search, "*") {
		left = true
	}
	if strings.HasSuffix(search, "*") {
		right = true
	}
	search = strings.Trim(search, "*")
	if left || right {
		// If not exact search, lowerCase
		ref = strings.ToLower(ref)
		search = strings.ToLower(search)
	}
	if left && right && !strings.Contains(ref, search) { // *part*
		return false
	} else if right && !left && !strings.HasPrefix(ref, search) { // start*
		return false
	} else if left && !right && !strings.HasSuffix(ref, search) { // *end
		return false
	} else if !left && !right && ref != search { // exact term
		return false
	}
	return true
}

File: frontend/assets/access.settings/res/js/AdminWorkspaces/model/DataSource.js
/*
 * Copyright 2007-2019 Charles du Jeu - Abstrium SAS <team (at) pyd.io>
 * This file is part of Pydio.
 *
 * Pydio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pydio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pydio.  If not, see <http://www.gnu.org/licenses/>.
 *
 * The latest code can be found at <https://pydio.com>.
 */

const Observable = require('pydio/lang/observable');
const PydioApi = require('pydio/http/api');
import LangUtils from 'pydio/util/lang'
import {ConfigServiceApi, HealthServiceApi, ObjectDataSource, ObjectEncryptionMode, ObjectStorageType,
    JobsServiceApi, RestUserJobRequest,EncryptionAdminListKeysRequest, RestListStorageBucketsRequest, RestCreateStorageBucketRequest} from 'cells-sdk';

class DataSource extends Observable {

    buildProxy(object){
        return new Proxy(object, {
            set:((target, p, value) => {
                let val = value;
                if(p === 'StorageType') {
                    const internal = target.StorageConfiguration && target.StorageConfiguration.cellsInternal=== 'true'
                    target.StorageConfiguration = internal?{cellsInternal: 'true'}:{};
                    if(val === 'LOCAL'){
                        target.StorageConfiguration = {"folder":"", "normalize":"false", ...target.StorageConfiguration}
                    } else if(val === 'S3'){
                        target.StorageConfiguration = {"customEndpoint":"", ...target.StorageConfiguration}
                    } else if(val === 'GCS') {
                        target.StorageConfiguration = {"jsonCredentials": "", ...target.StorageConfiguration}
                    }
                    this.internalInvalid = false;
                    target['ApiKey'] = target['ApiSecret'] = ''; // reset values
                } else if(p === 'Name') {
                    // Limit Name to 33 chars
                    val = LangUtils.computeStringSlug(val).replace("-", "").substr(0, 33);
                    if(this.existingNames && this.existingNames.indexOf(val) > -1) {
                        this.nameInvalid = true;
                    } else {
                        this.nameInvalid = false;
                    }
                } else if(p === 'folder') {
                    if (val[0] !== '/') {
                        val = '/' + val;
                    }
                } else if(p === 'invalid') {
                    this.internalInvalid = value;
                    this.notify('update');
                    return true;
                } else if (p === 'PeerAddress') {
                    if(value === 'ANY'){
                        val = '';
                    }
                }
                target[p] = val;
                this.notify('update');
                return true;
            }),
            get:((target, p) => {
                let out = target[p];
                if (out instanceof Array) {
                    return out;
                } else if (out instanceof Object) {
                    return this.buildProxy(out);
                } else if (p === 'StorageType') {
                    return out || 'LOCAL';
                } else if (p === 'PeerAddress') {
                    return out || 'ANY';
                } else {
                    return out;
                }
            })
        });
    }

    constructor(model, existingNames = [], createStructure = 'flat'){
        super();
        this.internalInvalid = false;
        this.nameInvalid = false;
        this.existingNames = existingNames;
        if (model) {
            this.model = model;
            if(!model.StorageConfiguration){
                model.StorageConfiguration = {};
            }
            this.snapshot = JSON.parse(JSON.stringify(model));
        } else {
            this.model = new ObjectDataSource();
            this.model.EncryptionMode = ObjectEncryptionMode.constructFromObject('CLEAR');
            this.model.StorageType = ObjectStorageType.constructFromObject('LOCAL');
            switch (createStructure) {
                case "flat":
                    this.model.StorageConfiguration = {"folder":"", "normalize":"false"};
                    this.model.FlatStorage = true;
                    break;
                case "structured":
                    this.model.StorageConfiguration = {"folder":"", "normalize":"false"};
                    this.model.FlatStorage = false;
                    break;
                case "internal":
                    this.model.StorageConfiguration = {"folder":"", "normalize":"false", "cellsInternal":"true"};
                    this.model.FlatStorage = true;
                    break;
            }
        }
        this.observableModel = this.buildProxy(this.model);

    }

    /**
     * @return {ObjectDataSource}
     */
    getModel(){
        return this.observableModel;
    }

    isValid(){
        if(this.internalInvalid || this.nameInvalid){
            return false;
        }
        if(this.model.StorageType === 'S3' || this.model.StorageType === 'AZURE') {
            return this.model.ApiKey && this.model.ApiSecret && this.model.Name && (this.model.ObjectsBucket || this.model.StorageConfiguration.bucketsRegexp);
        } else if(this.model.StorageType === 'GCS') {
            return this.model.Name && this.model.ObjectsBucket && this.model.StorageConfiguration && this.model.StorageConfiguration['jsonCredentials'];
        } else {
            return this.model.Name && this.model.StorageConfiguration && this.model.StorageConfiguration['folder'];
        }
    }

    /**
     *
     * @param translateFunc {Function} Translate function
     * @return {*}
     */
    getNameError(translateFunc){
        if(this.nameInvalid){
            return translateFunc('name.inuse')
        } else {
            return null;
        }
    }

    static loadDatasources(){
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        return api.listDataSources();
    }

    static loadVersioningPolicies(){
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        return api.listVersioningPolicies()
    }

    static loadStatuses(){
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        return api.listServices('STARTED');
    }

    static datasourceReady(dsName) {
        const api = new HealthServiceApi(PydioApi.getRestClient({silent: true}));
        return api.serviceReady(dsName)
    }

    static loadEncryptionKeys(){
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        return api.listEncryptionKeys(new EncryptionAdminListKeysRequest());
    }

    static loadBuckets(model, regexp = '') {
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        const request = new RestListStorageBucketsRequest();
        request.DataSource = model;
        if(regexp) {
            request.BucketsRegexp = regexp;
        }
        return api.listStorageBuckets(request);
    }

    static createBucket(model, bucketName) {
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        const request = new RestCreateStorageBucketRequest();
        request.DataSource = model;
        return api.createStorageBucket(bucketName, request)
    }

    deleteSource(){
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        return api.deleteDataSource(this.model.Name);
    }

    resyncSource(){
        const api = new JobsServiceApi(PydioApi.getRestClient());
        let req = new RestUserJobRequest();
        req.JobName = "datasource-resync";
        req.JsonParameters = JSON.stringify({dsName:this.model.Name});
        return api.userCreateJob("datasource-resync", req);
    }

    revert(){
        this.model = this.snapshot;
        this.observableModel = this.buildProxy(this.model);
        this.snapshot = JSON.parse(JSON.stringify(this.model));
        return this.observableModel;
    }

    saveSource(){
        const api = new ConfigServiceApi(PydioApi.getRestClient());
        return api.putDataSource(this.model.Name, this.model).then((res) => {
            this.snapshot = JSON.parse(JSON.stringify(this.model));
            this.notify('update');
        });
    }

    stripPrefix(data, prefix = ''){
        if(!prefix) {
            return data;
        }
        let obj = {};
        Object.keys(data).map((k) => {
            obj[k.replace(prefix, '')] = data[k];
        });
        return obj;

    }

    getDataWithPrefix(prefix = ''){
        if (prefix === '') {
            return this.model;
        }
        let data = {};
        Object.keys(this.model).forEach((k) => {
            data[prefix + k] = this.model[k];
            if(k === 'EncryptionMode' && !this.model[k]){
                data[prefix + k] = 'CLEAR';
            }
        });
        return data;
    }

}

export {DataSource as default};